build

```
@@ -164,6 +168,9 @@ def package_docker(config, target, src_dir, iteration, clean=False):
         fpacman = '-t pacman'
         fpm_args = fpm_args.replace('/usr/local', '/usr')
         shell('docker run --rm -v%s:/tgt -w/tgt %s %s' % (os.path.abspath('targets'), fpm_args, fpacman))
+    elif output == 'zip':
+        shell('rm -f targets/%s.zip' % archive)
+        shell('cd targets/%s/wkhtmltox && zip -r ../%s.zip *' % (target, archive))
```

```suggestion
        shell('cd targets/%s && zip -r ../%s.zip *' % (target, archive))
```
Can we make the change in the `after_compile` script instead? As this is specific to the target.

---

`/var/task/` and `/opt/` are the internals of AWS Lambda services logic. We can't affect them directly. We may put zip archive in S3 bucket, and tell Lambda to use it as a layer, in that case the contents of the archive will appear in `/opt`, or we may tell Lambda to use it as Lambda itself, in that case it will appear in `/var/task`. In both cases, we need to have `lib` directory of wkhtmltox placed directly in `/opt` or `/var/task`, otherwise it will not be used by OS. So, lib should be in the root of archive, and not a subdirectory of `wkhtmltox`. Otherwise, manual work is needed to use generated zip archive.

---

Ah, so the way it's generated is fine. Can we then do a `mv` inside the script so we can reuse the `zip` output without tying it to the lambda logic?

---

If we will execute the command suggested by you `cd targets/%s && zip -r ../%s.zip *`, it will generate the archive with the content of `targets/%s`, which is `app qt qt_configured wkhtmltox`. If we will rollback my last changes, then the command will be `cd targets/%s && zip -r ../%s.zip wkhtmltox`, and it will generate an archive with `wkhtmltox`, but `wkhtmltox` will be in the root of the archive, which is not suitable for lambda. 
Could you please give me a bit more details regarding your suggestion with `mv` in `after_compile` script?

---

With my last changes, the difference between `7z` and `zip` output, is that `7z` archives wkhtmltox directory and use it as root. `zip` archives wkhtmltox internals and put them in the root of the archive.

---

`shell('cd targets/%s/wkhtmltox && zip -r ../%s.zip *' % (target, archive))`

This does extract into `/opt` properly when using with lambda layers. :+1: for the changes.

---

Hmm, you're right -- it does make it complicated. Can we do a small tweak, then -- let's call it `output: lambda_zip` as this is specific to how lambdas are supposed to be? I don't want this to be confused with the other generic outputs (e.g. .tar or .7z).

---

Also, this has to be
```suggestion
        shell('cd targets/%s/wkhtmltox && zip -r ../../%s.zip *' % (target, archive))
```
as we're one level deeper in the folder hierarchy. I also think we need to add `targets/*.zip` in the Azure Pipelines configuration, as it'll generate but not upload otherwise.

---

@ashkulz thanks for your suggestions. The `lambda_zip` output name makes it clear that it's different than `7z` or `tar` and specific for Lambda output. I've fixed the azure pipeline as well.